include/openthread/srp_client.h

```diff
@@ -0,0 +1,503 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread SRP (Service Registration Protocol) client APIs.
+ */
+
+#ifndef OPENTHREAD_SRP_CLIENT_H_
+#define OPENTHREAD_SRP_CLIENT_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-srp
+ *
+ * @brief
+ *   This module includes functions that control SRP client behavior.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This structure represents a TXT record entry representing a key/value pair (RFC 6763 - section 6.3).
+ *
+ * The strings buffers pointed to by `mKey` and `mValue` MUST persist and remain unchanged after an instance of such a
+ * structure is passed to OpenThread (as part of `otSrpClientService` instance).
+ *
+ * An array of `otSrpTxtEntry` entries is used in `otSrpClientService` to specify the full TXT record (a list of
+ * entries).
+ *
+ */
+typedef struct otSrpTxtEntry
+{
+    /**
+     * The TXT record key string.
+     *
+     * If `mKey` is not NULL, then the entry is treated as key/value pair with `mValue` buffer providing the value.
+     *   - The entry is encoded as follows:
+     *        - A single string length byte followed by "key=value" format (without the quotation marks).
+              - In this case, the overall encoded length must be 255 bytes or less.
+     *   - If `mValue` is NULL, then key is treated as a boolean attribute and encoded as "key" (with no `=`).
+     *   - If `mValue` is not NULL but `mValueLength` is zero, then it is treated as empty value and encoded as "key=".
+     *
+     * If `mKey` is NULL, then `mValue` buffer is treated as an already encoded TXT-DATA and is appended as is in the
```

Comment:
It is just to provide one flexible way to pass in the TXT info, representing of *encoded* TXT-DATA blob, 

When preparing the message, it is verified that `mValue` (the encoded TXT-DATA) not null.

https://github.com/openthread/openthread/blob/220d040b1064c60440f9894cc7cad4bd30c003bc/src/core/net/srp_client.cpp#L52-L62

As for the encoded TXT-DATA mValue, the client might depends on the check by the server (like the others https://github.com/openthread/openthread/pull/6038#discussion_r554623667), and the server can silently ignore such invalid txt record.

Reply:
I can understand why we can use `mKey` for a special case in which `mValue` is treated as the raw TXT entry. However, I think such API protocol is not as simple and clear as it can be. 
I couldn't find any use case of `mValue` being the raw TXT entry in the current code. Maybe we can introduce this feature only when we find it beneficial so as not to increase code size.

Thoughts?

Reply:
This may already be clear to you, but just to be sure, what it means for mValue to be non-null and mKey to be null is that mValue is already in the form "key=value" or "key". So it's pretty easy to implement, unless you want to check it for syntactic correctness, which I think is unnecessary and doesn't match the philosophy of insisting that the caller not pass invalid data.

Reply:
This is intended to provide flexibility in the API (be able to hook OT API to different users following different models for representing TXT info).  The behaviors are explained in detail in the documentation of `otSrpTxtEntry` definition.

The support for already encoded model actually does not add much code overhead and is very simple to add (as @Abhayakara also mentioned).